clients/tfchain-client-go/impl.go

```diff
@@ -37,145 +36,437 @@ type Versioned struct {
 type Conn = *gsrpc.SubstrateAPI
 type Meta = *types.Metadata
 
+// Pool connection
+type poolConn struct {
+	conn     Conn
+	meta     Meta
+	url      string
+	lastUsed atomic.Int64 // Unix timestamp
+	inUse    atomic.Bool
+}
+
+func (pc *poolConn) isHealthy() bool {
+	if pc == nil || pc.conn == nil || pc.meta == nil {
+		return false
+	}
+	_, err := getTime(pc.conn, pc.meta)
+	return err == nil
+}
+
+func (pc *poolConn) close() {
+	if pc != nil && pc.conn != nil {
+		pc.conn.Client.Close()
+		pc.conn = nil
+		pc.meta = nil
+		log.Debug().Str("url", pc.url).Msg("closed connection")
+	}
+}
+
 type Manager interface {
+	GetConnection(ctx context.Context) (*Substrate, error)
+	Close() error
+
+	// Deprecated methods
 	Raw() (Conn, Meta, error)
 	Substrate() (*Substrate, error)
 }
 
-type mgrImpl struct {
-	urls []string
+type ManagerConfig struct {
+	// Maximum number of connections in the pool
+	MaxPoolSize int
+	// Minimum number of connections to maintain
+	MinPoolSize int
+	// Maximum time a connection can be idle before being closed (if the pool has more than MinPoolSize)
+	MaxIdleTime time.Duration
+	// Interval between health checks
+	// After thinking about it, we don't need to periodically check the health of the connections
+	// because this creates a lot of overhead
+	// so instead we just check the health when we need to and do the maintanance in demand
+	// HealthCheckInterval time.Duration
+	// Timeout for creating new connections
+	ConnectionTimeout time.Duration
```

Comment:
Typically, this is the behavior you want.
Mainly to avoid waiting indefinitely on a stuck connection instead of attempting to reconnect.
Also, this parameter is not used at all and should be cleaned up.
The reason I removed the timeout logic (not completely as you noticed) is that I found it already implemented in the inner substrate client out of the box, so it wasn't needed in our code.